scio-parquet/src/main/scala/com/spotify/scio/parquet/read/ParquetReadFn.scala

```diff
@@ -0,0 +1,237 @@
+    val options = HadoopReadOptions.builder(conf.get()).build
+    ParquetFileReader.open(BeamInputFile.of(file.openSeekable), options)
+  }
+
+  @GetRestrictionCoder def getRestrictionCoder = new OffsetRange.Coder
+
+  @GetInitialRestriction
+  def getInitialRestriction(@Element file: ReadableFile): OffsetRange = {
+    val reader = parquetFileReader(file)
+
+    val offsetRangeResult = Try(new OffsetRange(0, reader.getRowGroups.size)).toEither
+    reader.close()
+
+    offsetRangeResult.fold(throw _, identity)
+  }
+
+  @NewTracker def newTracker(@Restriction restriction: OffsetRange) =
+    new OffsetRangeTracker(restriction)
+
+  @GetSize
+  def getSize(@Element file: ReadableFile, @Restriction restriction: OffsetRange): Double = {
+    val (_, size) = getRecordCountAndSize(file, restriction)
+    size
+  }
+
+  @SplitRestriction
+  def split(
+    @Restriction restriction: OffsetRange,
+    output: DoFn.OutputReceiver[OffsetRange],
+    @Element file: ReadableFile
+  ): Unit = {
+    val reader = parquetFileReader(file)
+
+    val splitResult = Try {
+      splitRowGroupsWithLimit(
+        restriction.getFrom,
+        restriction.getTo,
+        reader.getRowGroups.asScala.toSeq
+      ).foreach(output.output)
+    }.toEither
+
+    reader.close()
+
+    splitResult.fold(throw _, identity)
+  }
+
+  @ProcessElement
+  def processElement(
+    @Element file: ReadableFile,
+    tracker: RestrictionTracker[OffsetRange, Long],
+    outputReceiver: DoFn.OutputReceiver[R]
+  ): Unit = {
+    logger.debug(
+      "start {} to {}",
+      tracker.currentRestriction.getFrom,
+      tracker.currentRestriction.getTo
+    )
+    val options: ParquetReadOptions = HadoopReadOptions.builder(conf.get()).build
+
+    val reader = ParquetFileReader.open(BeamInputFile.of(file.openSeekable), options)
+    try {
+      val filter = options.getRecordFilter
+      val hadoopConf = options.asInstanceOf[HadoopReadOptions].getConf
+      val parquetFileMetadata = reader.getFooter.getFileMetaData
+      val fileSchema = parquetFileMetadata.getSchema
+      val fileMetadata = parquetFileMetadata.getKeyValueMetaData
+      val readSupport = readSupportFactory.readSupport
+
+      val readContext = readSupport.init(
+        new InitContext(
+          hadoopConf,
+          fileMetadata.asScala.map { case (k, v) =>
+            k -> ImmutableSet.of(v).asInstanceOf[JSet[String]]
+          }.asJava,
+          fileSchema
+        )
+      )
+      reader.setRequestedSchema(readContext.getRequestedSchema)
+
+      val columnIOFactory = new ColumnIOFactory(parquetFileMetadata.getCreatedBy)
+      val recordConverter =
+        readSupport.prepareForRead(hadoopConf, fileMetadata, fileSchema, readContext)
+      val columnIO = columnIOFactory.getColumnIO(readContext.getRequestedSchema, fileSchema, true)
+
+      var currentRowGroupIndex = tracker.currentRestriction.getFrom
+      (0L until currentRowGroupIndex).foreach(_ => reader.skipNextRowGroup())
```

Comment:
Depending on the underlying file storage, seeking can be quite an expensive operation. For example seeking on GCS literally means sending a new HTTP request to the GCS API. Not sure if it's possible but ideally it would be better to directly "jump" to the desired row group.

Reply:
yeah, that's true, but there is no API for jumping to a row group directly or at least I didn't find it. This shouldn't be a big problem because in Spotify we try to set row group size almost equal to the shard size so the result shards have only one row group per file.